src/sentry/releases/servicer.py

```diff
@@ -0,0 +1,128 @@
+            organization=organization,
+            time_range=TimeRange(start=start, end=end),
+            project_slugs=project_slugs,
+            versions=versions,
+            environments=environments,
+        ):
+            project = release_thresholds.project
+            release = release_thresholds.release
+            for i, threshold in enumerate(release_thresholds.thresholds):
+                if threshold.threshold_type not in thresholds_by_type:
+                    thresholds_by_type[threshold.threshold_type] = {
+                        "project_ids": [],
+                        "releases": [],
+                        "thresholds": [],
+                    }
+                thresholds_by_type[threshold.threshold_type]["project_ids"].append(project.id)
+                thresholds_by_type[threshold.threshold_type]["releases"].append(release.version)
+                if threshold.threshold_type not in query_windows_by_type:
+                    query_windows_by_type[threshold.threshold_type] = {
+                        "start": datetime.now(tz=timezone.utc),
+                        "end": datetime.now(tz=timezone.utc),
+                    }
+
+                latest_deploy: Deploy | None = release_thresholds.get_latest_deploy_id_by_threshold(
+                    index=i
+                )
+                # NOTE: query window starts at the earliest release up until the latest threshold window
+                if latest_deploy:
+                    threshold_start = latest_deploy.date_finished
+                else:
+                    threshold_start = release.date
+
+                query_windows_by_type[threshold.threshold_type]["start"] = min(
+                    threshold_start, query_windows_by_type[threshold.threshold_type]["start"]
+                )
+                query_windows_by_type[threshold.threshold_type]["end"] = max(
+                    threshold_start + timedelta(seconds=threshold.window_in_seconds),
+                    query_windows_by_type[threshold.threshold_type]["end"],
+                )
+                # NOTE: enriched threshold is SERIALIZED
+                # meaning project and environment models are dictionaries
+                enriched_threshold: EnrichedThreshold = serialize(threshold)
+                # NOTE: start/end for a threshold are different from start/end for querying data
+                enriched_threshold.update(
+                    {
+                        "key": release_thresholds.get_threshold_key(),
+                        "start": threshold_start,
+                        "end": threshold_start
+                        + timedelta(
+                            seconds=threshold.window_in_seconds
+                        ),  # start + threshold.window
+                        "release": release.version,
+                        "project_slug": project.slug,
+                        "project_id": project.id,
+                        "is_healthy": False,
```

Comment:
why isn't this set at the time of creation of this object? Why do we have to loop back through the lists again to calculate the value?

Reply:
not quite sure i understand the question...

Here we're just initializing the value. we haven't calculated anything yet.

we absolutely _could_ just do all the calculations here too?
But separation of concerns 🤷 

O(2N) isn't any more efficient than O(N). This makes it easier to understand the logical steps being taken